internals-js/src/directiveAndTypeSpecification.ts

```diff
@@ -334,3 +336,19 @@ function ensureSameArguments(
   return errors;
 }
 
+function isValidInputTypeRedefinition(expectedType: InputType, actualType: InputType): boolean {
+  // If the expected type is a custom scalar, then we allow the redefinition to be another type (unless it's a custom scalar, in which
+  // case it has to be the same scalar). The rational being that since graphQL does no validation of values passed to a custom scalar,
+  // any code that gets some value as input for a custom scalar has to do validation manually, and so there is little harm in allowing
+  // a redefinition with another type since any truly invalid value would failed that "manual validation". In practice, this leeway
+  // make sense because many scalar will tend to accept only one kind of values (say, strings) and exists only to inform that said string
+  // needs to follow a specific format, and in such case, letting user redefine the type as String adds flexibility while doing little harm.
+  if (isListType(expectedType)) {
+    return isListType(actualType) && isValidInputTypeRedefinition(expectedType.ofType, actualType.ofType);
+  }
+  if (isNonNullType(expectedType)) {
```

Comment:
I didn't change this one as I'll admit kind of disliking having a `else` (not a strong dislike obviously, but I frown when I see one). I guess my main pseudo-objective rational is that I sometimes like `ifs-that-return` as a mean to avoid creating too much indentation (which is in and of itself a bit of a subjective goal in the first place), which only work if you don't have a `else`.

Anyway, overall a subjective preference on my part, and I can change my preference if you feel a `else` makes a marked difference for you, but I read your comment as a weak preference, so wanted to justify ignoring it.

Reply:
Very weak preference. I mostly bring it up to have this discussion so that I can adopt the right style for my code too. Thanks for the explanation.